File: Python/can_you_reach_the_exit.py
def path_finder(maze):
    maze_arr = generate_maze(maze)
    print(doDFS(maze_arr))


def generate_maze(maze_str):
    grid = []
    maze_str_split = maze_str.split("\n")

    for small_grid in maze_str_split:
        sub_grid = []
        for cell in small_grid:
            sub_grid.append(cell)
        grid.append(sub_grid)

    return grid


class Node:
    def __init__(self, row, col, val, graph):
        self.row = row
        self.col = col
        self.val = val
        self.graph = graph
        self.neighbours = []

    def generate_neighbours(self):
        if self.col + 1 < len(self.graph):
            self.neighbours.append(
                Node(self.row, self.col + 1, self.graph[self.row][self.col + 1], self.graph))

        if self.col - 1 > -1:
            self.neighbours.append(
                Node(self.row, self.col - 1, self.graph[self.row][self.col], self.graph))

        if self.row + 1 < len(self.graph):
            self.neighbours.append(
                Node(self.row + 1, self.col, self.graph[self.row + 1][self.col], self.graph))

        if self.row - 1 > -1:
            self.neighbours.append(
                Node(self.row - 1, self.col, self.graph[self.row - 1][self.col], self.graph))


def toString(node: Node):
    print("Row: ", node.row)
    print("Col: ", node.col)
    print("Val: ", node.val)
    print("Neighbours:", end='')
    for neighbour in node.neighbours:
        print("\t[", end='')
        print("Row: ", neighbour.row, end='')
        print(" Col: ", neighbour.col, end='')
        print(" Val: ", neighbour.val, end='')
        print("]")


def doDFS(grid):
    start_node = Node(0, 0, grid[0][0], grid)
    stack = []
    stack.append(start_node)
    found_path = False
    visited = {}

    while len(stack) > 0:
        current = stack.pop()

        if current.row == len(grid) - 1 and current.col == len(grid) - 1:
            found_path = True
            break

        if not visited[str(Node.id)]:
            visited[Node.id] = True
            Node.id += 1
            current.generate_neighbours()
            toString(current)
            for neighbour in current.neighbours:
                if neighbour.val == '.':
                    stack.append(neighbour)

    return found_path


a = "\n".join([
    ".W.",
    ".W.",
    "..."
])

# print(generate_maze(a))

b = "\n".join([
    ".W.",
    ".W.",
    "W.."
])

c = "\n".join([
    "......",
    "......",
    "......",
    "......",
    "......",
    "......"
])

d = "\n".join([
    "......",
    "......",
    "......",
    "......",
    ".....W",
    "....W."
])

print(path_finder(a), True)
# print(path_finder(b), False)
# print(path_finder(c), True)
# print(path_finder(d), False)

File: JavaScript/can_you_reach_the_exit.js
/* 
Task
You are at position [0, 0] in maze NxN and you can only move in one of the four cardinal directions (i.e. North, East, South, West). Return true if you can reach position [N-1, N-1] or false otherwise.

Empty positions are marked .. Walls are marked W. Start and exit positions are empty in all test cases. 
*/

/* 
    [[0, 1 , 2],
     [0, 1 , 2],
     [0, 1 , 2],
     [0, 1 , 2],
     [0, 1 , 2]
    ]

    i - 1, j
    i + 1, j
    i, j - 1
    i, j + 1

*/

// NOT WORKING.......
//TODO: FIX IT.....

// function pathFinder(maze) {
//   const grid_arr = createGridArr(maze);
//   return doDFS(grid_arr);
// }

// function createGridArr(string) {
//   const string_arr = string.split("\n");
//   const grid = [];
//   for (let i = 0; i < string_arr.length; i++) {
//     const cells_arr = string_arr[i].trim();
//     const cell = [];
//     for (let j = 0; j < cells_arr.length; j++) {
//       cell.push(cells_arr[j]);
//     }
//     grid.push(cell);
//   }
//   return grid;
// }

// class Node {
//   constructor(row, col, grid) {
//     this.row = row;
//     this.col = col;
//     this.grid = grid;
//     this.value = grid[this.row][this.col];
//     this.getNeighbours = function () {
//       const neighbours = [];
//       if (this.col - 1 > -1) {
//         neighbours.push(new Node(this.row, this.col - 1, grid));
//       }

//       if (this.col + 1 < this.grid.length) {
//         neighbours.push(new Node(this.row, this.col + 1, grid));
//       }

//       if (this.row - 1 > -1) {
//         neighbours.push(new Node(this.row - 1, this.col, grid));
//       }

//       if (this.row + 1 < this.grid.length) {
//         neighbours.push(new Node(this.row + 1, this.col, grid));
//       }

//       return neighbours;
//     };
//   }
// }

// function doDFS(grid) {
//   const start = new Node(0, 0, grid);
//   const stack = [];
//   stack.push(start);
//   const visited = {};

//   while (stack.length > 0) {
//     const current = stack.pop();
//     if (current.row == grid.length - 1 && current.col == grid.length - 1) {
//       return true;
//     }
//     if (visited[current] == undefined) {
//       visited[current] = true;
//       const current_neighbours = current.getNeighbours();
//       for (let i = 0; i < current_neighbours.length; i++) {
//         const current_neighbour = current_neighbours[i];
//         if (current_neighbour.value != "W") {
//           stack.push(current_neighbour);
//         }
//       }
//     }
//   }
//   return false;
// }

// function pathFinder(maze) {
//   const grid = generateMazeGrid(maze);
//   const startNode = new Node(0, 0, grid);
//   const queue = [];
//   startNode.visited = true;
//   queue.push(startNode);

//   while (queue.length > 0) {
//     const current = queue.shift();
//     if (current.row == grid.length - 1 && current.col == grid.length - 1) {
//       return true;
//     }
//     const neighbours = current.generateNeighbours();
//     for (let i = 0; i < neighbours.length; i++) {
//       const neighbour = neighbours[i];
//       if (!neighbour.visited) {
//         neighbour.visited = true;
//         if (neighbour.val != "W") {
//           queue.push(neighbour);
//         }
//       }
//     }
//   }
//   return false;
// }

// function generateMazeGrid(maze) {
//   const grid = [];
//   const subgrid = maze.split("\n");

//   for (let i = 0; i < subgrid.length; i++) {
//     const cells = subgrid[i].trim();
//     const subgridArr = [];
//     for (let j = 0; j < cells.length; j++) {
//       subgridArr.push(cells[j]);
//     }
//     grid.push(subgridArr);
//   }

//   return grid;
// }

// class Node {
//   constructor(row, col, grid) {
//     this.row = row;
//     this.col = col;
//     this.grid = grid;
//     this.val = grid[row][col];
//     this.visited = false;
//     this.generateNeighbours = function () {
//       const neighbours = [];
//       if (this.row + 1 < grid.length) {
//         neighbours.push(new Node(this.row + 1, this.col, this.grid));
//       }

//       if (this.row - 1 > -1) {
//         neighbours.push(new Node(this.row - 1, this.col, this.grid));
//       }

//       if (this.col + 1 < grid.length) {
//         neighbours.push(new Node(this.row, this.col + 1, this.grid));
//       }

//       if (this.col - 1 > -1) {
//         neighbours.push(new Node(this.row, this.col - 1, this.grid));
//       }
//       return neighbours;
//     };
//   }
// }

function pathFinder(maze) {
  const grid = createGrid(maze);
  const graph = [];

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid.length; j++) {
      const node = new Node(i, j, grid.length);
      node.setSymbol(grid[i][j]);
      graph.push(node);
    }
  }

  graph.forEach((node) => {
    node.getNeighbours();
  });

  const start = graph[0];
  const queue = [];
  start.visited = true;
  queue.push(start);

  while (queue.length > 0) {
    const current = queue.shift();
    if (current.row == grid.length - 1 && current.col == grid.length - 1) {
      return true;
    }
    for (let i = 0; i < current.neighbours.length; i++) {
      const neighbour = current.neighbours[i];
      if (!neighbour.visited && neighbour.symbol != "W") {
        neighbour.visited = true;
        queue.push(neighbour);
      }
    }
  }
  return false;
}

class Node {
  constructor(row, col, size) {
    this.row = row;
    this.col = col;
    this.size = size;
    this.visited = false;
    this.neighbours = [];
  }

  getNeighbours() {
    if (this.col + 1 < this.size)
      this.neighbours.push(new Node(this.row, this.col + 1, this.size));

    if (this.row + 1 < this.size)
      this.neighbours.push(new Node(this.row + 1, this.col, this.size));

    if (this.col - 1 > -1)
      this.neighbours.push(new Node(this.row, this.col - 1, this.size));

    if (this.row - 1 > -1)
      this.neighbours.push(new Node(this.row - 1, this.col, this.size));
  }

  setSymbol(symbol) {
    this.symbol = symbol;
  }
}

function createGrid(maze) {
  const grid = [];
  const rows = maze.split("\n");
  const size = rows.length;

  for (let i = 0; i < rows.length; i++) {
    let row = rows[i];
    row = row.trim();
    const gridRow = [];
    for (let j = 0; j < row.length; j++) {
      gridRow.push(row[j]);
    }
    grid.push(gridRow);
  }
  return grid;
}

// console.log(
//   pathFinder(`.W.
//     .W.
//     ...`),
//   true
// );

// console.log(
//   pathFinder(`......
//        ......
//        ......
//        ......
//        ......
//        ......`),
//   true
// );

// console.log(pathFinder(".W.\n.W.\nW.."));

// console.log(
//   pathFinder(
//     `.W.
//     .W.
//     ...`
//   ),
//   true
// );

console.log(
  pathFinder(
    `.W.
    .W.
    W..`
  ),
  false
);

// console.log(
//   pathFinder(
//     `......
//     ......
//     ......
//     ......
//     ......
//     ......`
//   ),
//   true
// );

console.log(
  pathFinder(
    `......
    ......
    ......
    ......
    .....W
    ....W.`
  ),
  false
);
